internal/util/telemetry/reporter_test.go

```diff
@@ -88,3 +93,62 @@ func TestNewReporterLock(t *testing.T) {
 		})
 	}
 }
+
+func TestReporterReport(t *testing.T) {
+	t.Parallel()
+
+	for name, tc := range map[string]struct {
+		f                *Flag
+		reporterResponse string
+		latestVersion    string
+		updateAvailable  bool
+	}{
+		"UpdateAvailable": {
+			f:                &Flag{v: pointer.ToBool(true)},
+			reporterResponse: `{"update_available": true, "latest_version": "0.3.4"}`,
+			latestVersion:    "0.3.4",
+			updateAvailable:  true,
+		},
+		"UpdateUnavailable": {
+			f:                &Flag{v: pointer.ToBool(true)},
+			reporterResponse: `{"update_available": false, "latest_version": "0.3.4"}`,
+			latestVersion:    "",
+			updateAvailable:  false,
+		},
+	} {
+		name, tc := name, tc
+
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			// use httptest.NewServer so http request on ts.URL returns tc.reporterResponse body
+			ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusCreated)
+				fmt.Fprintln(w, tc.reporterResponse)
+			}))
+			defer ts.Close()
+
+			provider, err := state.NewProvider("")
+			require.NoError(t, err)
+
+			opts := NewReporterOpts{
+				URL:           ts.URL,
+				F:             tc.f,
+				ConnMetrics:   connmetrics.NewListenerMetrics().ConnMetrics,
+				P:             provider,
+				L:             zap.L(),
+				ReportTimeout: 1 * time.Minute,
+			}
+
+			r, err := NewReporter(&opts)
+			assert.NoError(t, err)
+
+			r.report(testutil.Ctx(t))
+
+			s := r.P.Get()
```

Comment:
This is the part I don't understand. I see `r.report` is called above, and the info on wether the update is available is identified as a part of `r.report`. Beacon returns `updateAvailable` only if `POST` was requested - https://github.com/FerretDB/beacon/blob/main/internal/handler.go#L156. With that, I don't understand what do we expect with this `r.P.Get()` as Beacon won't return `updateAvailable` for it.

Reply:
`r.report()` either calls `POST` or it doesn't call telemetry at all. `GET` is not called by `r.report` I guess it's used by something else. When `POST is called, it updates the state of the provider by `r.P.Update` if any update is available.

To check if the state is updated by `r.report()`, we use provider to get value `r.P.Get()`.

When beacon doesn't call `POST` indeed there won't be any update available.

Added comments in the code, let me know if that clarifies :hugs:

Reply:
Meanwhile, I can update the reporter to return the state `(r *Reporter) report(ctx context.Context) *state.State` which removes the need to call provider from test.

Reply:
Ah, ok! I think I now understand :)